commands/server_test.go

```diff
@@ -58,10 +58,14 @@ func TestServer(t *testing.T) {
 		c.Assert(err, qt.IsNil)
 	}()
 
-	// There is no way to know exactly when the server is ready for connections.
-	// We could improve by something like https://golang.org/pkg/net/http/httptest/#Server
-	// But for now, let us sleep and pray!
-	time.Sleep(2 * time.Second)
+	// Wait for the server to become ready or fail on timeout
+	k := time.NewTimer(5 * time.Second)
+	select {
+	case <-k.C:
+		t.Error("hugo server took too long to respond")
+	case <-scmd.ready:
```

Comment:
OK, I have read through the code;  I don't see how this detects _when the server is ready for connections_?

Reply:
When the `hugo server` command is executed, it blocks at [this `if` statement](https://github.com/gohugoio/hugo/blob/0303c29b63c8a4b8a1b515251c1b618196a1174a/commands/server.go#L548-L555) until it can receive from either the `stop` channel or `sigs`. In this PR, the server [sends to `ready`](https://github.com/gohugoio/hugo/blob/0303c29b63c8a4b8a1b515251c1b618196a1174a/commands/server.go#L546) just before the `serve` function blocks. I put the `s.ready <- struct{}{}` statement just before the command receives from the `sigs`/`stop` channels to ensure that all setup work completed by `serve` had finished.

Reply:
* The servers (one or more) are started in their own go routine.
* The printing of "Press Ctrl+C to stop" may come before the server(s) are started.
* I don't know of any way to detect "servers are ready to receive connections", and if it was, I don't see how this PR hooks into the `http` package.

Reply:
It was a misjudgment on my part to conflate the `hugo server` command finishing its setup work with the HTTP servers being ready for requests. 

One solution is to add a lightweight `/ready` endpoint to each HTTP server, probably in `*fileServer.createEndpoint`. This would simply return `200` to any GET request. We could then use a test helper function to repeatedly—and synchronously—poll the readiness endpoint of each HTTP server on a tight loop (say, every few milliseconds). The helper would return `nil` on success (and the test that calls it would continue) or cause the test to fail after a timeout. 

While this would still require sleeping between retries, it would be more deterministic than using a single `time.Sleep` call, and would probably save time. I would just need to figure out how to cleanly pass the base URL of each server to the test helper.

What do you think?

Reply:
> it also means that any new hugo server test would add another two seconds to the test suite. 

I suspect the starting point for this issue is a little bit flawed: The above statement isn't true; we can 1) Run server tests in parallel (each server gets its own port) and 2) Run many test case for each server start. All of Hugo's tests have been running ... thousands of times, and I don't remember one time the server tests have failed (at least not in recent times).

This should be supported in the Go stdlib, adding to the complexity here isn't worth it.